// @oruga-ui/oruga-next/dist/esm/table.mjs
import { defineComponent, resolveComponent, openBlock, createBlock, createVNode, withCtx, withDirectives, toDisplayString, vShow, createCommentVNode, Fragment, renderList, h, renderSlot, createApp, mergeProps, withKeys, withModifiers, createTextVNode, capitalize, Transition } from 'vue';
import { getValueByPath, toCssDimension, indexOf, debounce, escapeRegExpChars, removeDiacriticsFromString } from './helpers.mjs';
import { getOptions } from './config.mjs';
import { B as BaseComponentMixin, b as registerComponent } from './plugins-ba5e2464.mjs';
import { s as script$6 } from './Icon-a7b85e86.mjs';
import './FormElementMixin-29581863.mjs';
import { s as script$a } from './Input-19e906a1.mjs';
import { s as script$4 } from './Button-c75f7992.mjs';
import './CheckRadioMixin-f5b57344.mjs';
import { s as script$9 } from './Checkbox-8670724c.mjs';
import { M as MatchMediaMixin } from './MatchMediaMixin-09b141ca.mjs';
import { s as script$7 } from './Field-937b4591.mjs';
import { s as script$5 } from './Select-2cca695f.mjs';
import './ssr-85c76d50.mjs';
import { s as script$b } from './Loading-20d29914.mjs';
import { s as script$8 } from './Pagination-02a660a8.mjs';
import { S as SlotComponent } from './SlotComponent-11a269f3.mjs';

var script = defineComponent({
    name: 'OTableMobileSort',
    components: {
        [script$4.name]: script$4,
        [script$5.name]: script$5,
        [script$6.name]: script$6,
        [script$7.name]: script$7
    },
    inject: ['$table'],
    emits: ['sort'],
    props: {
        currentSortColumn: Object,
        columns: Array,
        placeholder: String,
        iconPack: String,
        sortIcon: {
            type: String,
            default: 'arrow-up'
        },
        sortIconSize: {
            type: String,
            default: 'small'
        },
        isAsc: Boolean
    },
    data() {
        return {
            mobileSort: getValueByPath(this.currentSortColumn, 'newKey'),
            defaultEvent: {
                shiftKey: true,
                altKey: true,
                ctrlKey: true
            },
            ignoreSort: false
        };
    },
    computed: {
        showPlaceholder() {
            return !this.columns || !this.columns.some((column) => (getValueByPath(column, 'newKey') === this.mobileSort));
        },
        sortableColumns() {
            if (!this.columns)
                return [];
            return this.columns.filter(c => c.sortable);
        },
        isCurrentSort() {
            return getValueByPath(this.currentSortColumn, 'newKey') === this.mobileSort;
        }
    },
    watch: {
        mobileSort(value) {
            if (this.currentSortColumn.newKey === value)
                return;
            const column = this.sortableColumns.filter(c => (getValueByPath(c, 'newKey') === value))[0];
            this.$emit('sort', column, this.defaultEvent);
        },
        currentSortColumn(column) {
            this.mobileSort = getValueByPath(column, 'newKey');
        }
    },
    methods: {
        sort() {
            const column = this.sortableColumns.filter(c => (getValueByPath(c, 'newKey') === this.mobileSort))[0];
            this.$emit('sort', column, this.defaultEvent);
        }
    }
});

function render(_ctx, _cache, $props, $setup, $data, $options) {
  const _component_o_select = resolveComponent("o-select");

  const _component_o_icon = resolveComponent("o-icon");

  const _component_o_button = resolveComponent("o-button");

  const _component_o_field = resolveComponent("o-field");

  return openBlock(), createBlock("div", {
    class: _ctx.$table.mobileSortClasses
  }, [createVNode(_component_o_field, null, {
    default: withCtx(() => [createVNode(_component_o_select, {
      modelValue: _ctx.mobileSort,
      "onUpdate:modelValue": _cache[1] || (_cache[1] = $event => _ctx.mobileSort = $event),
      expanded: ""
    }, {
      default: withCtx(() => [_ctx.placeholder ? withDirectives((openBlock(), createBlock("option", {
        key: 0,
        value: {},
        selected: "",
        disabled: "",
        hidden: ""
      }, toDisplayString(_ctx.placeholder), 513
      /* TEXT, NEED_PATCH */
      )), [[vShow, _ctx.showPlaceholder]]) : createCommentVNode("v-if", true), (openBlock(true), createBlock(Fragment, null, renderList(_ctx.sortableColumns, (column, index) => {
        return openBlock(), createBlock("option", {
          key: index,
          value: column.newKey
        }, toDisplayString(column.label), 9
        /* TEXT, PROPS */
        , ["value"]);
      }), 128
      /* KEYED_FRAGMENT */
      ))]),
      _: 1
    }, 8
    /* PROPS */
    , ["modelValue"]), createVNode(_component_o_button, {
      onClick: _ctx.sort
    }, {
      default: withCtx(() => [withDirectives(createVNode(_component_o_icon, {
        icon: _ctx.sortIcon,
        pack: _ctx.iconPack,
        size: _ctx.sortIconSize,
        both: "",
        rotation: !_ctx.isAsc ? 180 : 0
      }, null, 8
      /* PROPS */
      , ["icon", "pack", "size", "rotation"]), [[vShow, _ctx.isCurrentSort]])]),
      _: 1
    }, 8
    /* PROPS */
    , ["onClick"])]),
    _: 1
  })], 2
  /* CLASS */
  );
}

script.render = render;
script.__file = "src/components/table/TableMobileSort.vue";

/**
 * @displayName Table Column
 */
var script$1 = defineComponent({
    name: 'OTableColumn',
    inject: ['$table'],
    props: {
        label: String,
        customKey: [String, Number],
        field: String,
        meta: [String, Number, Boolean, Function, Object, Array],
        width: [Number, String],
        numeric: Boolean,
        /**
         * Optional, position of column content
         * @values left, centered, right
         */
        position: {
            type: String,
            validator(value) {
                return [
                    'left',
                    'centered',
                    'right'
                ].indexOf(value) > -1;
            }
        },
        searchable: Boolean,
        sortable: Boolean,
        visible: {
            type: Boolean,
            default: true
        },
        customSort: Function,
        customSearch: Function,
        sticky: Boolean,
        headerSelectable: Boolean,
        /** Adds native attributes to th :th-attrs="(column)" => ({})" */
        thAttrs: {
            type: Function,
            default: () => ({})
        },
        /** Adds native attributes to td :td-attrs="(row, column)" => ({})" */
        tdAttrs: {
            type: Function,
            default: () => ({})
        },
        subheading: String
    },
    data() {
        return {
            newKey: undefined,
            thAttrsData: {},
            tdAttrsData: []
        };
    },
    computed: {
        style() {
            return {
                width: toCssDimension(this.width)
            };
        },
        hasDefaultSlot() {
            return this.$slots.default;
        },
        hasSearchableSlot() {
            return this.$slots.searchable;
        },
        hasHeaderSlot() {
            return this.$slots.header;
        },
        isHeaderUnselectable() {
            return !this.headerSelectable && this.sortable;
        }
    },
    created() {
        if (!this.$table) {
            throw new Error('You should wrap oTableColumn on a oTable');
        }
        this.newKey = this.$table._nextSequence();
        this.$table._addColumn(this);
    },
    beforeMount() {
        if (typeof this.thAttrs !== 'undefined') {
            this.thAttrsData = this.thAttrs(this);
        }
    },
    beforeUnmount() {
        this.$table._removeColumn(this);
    },
    render() {
        // renderless
        return h('span', { 'data-id': this.newKey }, this.label);
    }
});

script$1.__file = "src/components/table/TableColumn.vue";

var script$2 = defineComponent({
  name: 'OTablePagination',
  components: {
    [script$8.name]: script$8
  },
  emits: ['update:currentPage', 'page-change'],
  props: {
    paginated: Boolean,
    total: [Number, String],
    perPage: [Number, String],
    currentPage: [Number, String],
    paginationSimple: Boolean,
    paginationSize: String,
    rounded: Boolean,
    iconPack: String,
    rootClass: [String, Array, Object],
    ariaNextLabel: String,
    ariaPreviousLabel: String,
    ariaPageLabel: String,
    ariaCurrentLabel: String
  },

  data() {
    return {
      newCurrentPage: this.currentPage
    };
  },

  watch: {
    currentPage(newVal) {
      this.newCurrentPage = newVal;
    }

  },
  methods: {
    /**
    * Paginator change listener.
    */
    pageChanged(page) {
      this.newCurrentPage = page > 0 ? page : 1;
      this.$emit('update:currentPage', this.newCurrentPage);
      this.$emit('page-change', this.newCurrentPage);
    }

  }
});

const _hoisted_1 = {
  key: 0
};
function render$1(_ctx, _cache, $props, $setup, $data, $options) {
  const _component_o_pagination = resolveComponent("o-pagination");

  return openBlock(), createBlock("div", {
    class: _ctx.rootClass
  }, [createVNode("div", null, [renderSlot(_ctx.$slots, "default")]), createVNode("div", null, [_ctx.paginated ? (openBlock(), createBlock("div", _hoisted_1, [createVNode(_component_o_pagination, {
    "icon-pack": _ctx.iconPack,
    total: _ctx.total,
    "per-page": _ctx.perPage,
    simple: _ctx.paginationSimple,
    size: _ctx.paginationSize,
    current: _ctx.newCurrentPage,
    rounded: _ctx.rounded,
    onChange: _ctx.pageChanged,
    "aria-next-label": _ctx.ariaNextLabel,
    "aria-previous-label": _ctx.ariaPreviousLabel,
    "aria-page-label": _ctx.ariaPageLabel,
    "aria-current-label": _ctx.ariaCurrentLabel
  }, null, 8
  /* PROPS */
  , ["icon-pack", "total", "per-page", "simple", "size", "current", "rounded", "onChange", "aria-next-label", "aria-previous-label", "aria-page-label", "aria-current-label"])])) : createCommentVNode("v-if", true)])], 2
  /* CLASS */
  );
}

script$2.render = render$1;
script$2.__file = "src/components/table/TablePagination.vue";

/**
 * Tabulated data are sometimes needed, it's even better when it's responsive
 * @displayName Table
 * @requires ./TableColumn.vue
 * @style _table.scss
 */
var script$3 = defineComponent({
    name: 'OTable',
    components: {
        [script$4.name]: script$4,
        [script$9.name]: script$9,
        [script$6.name]: script$6,
        [script$a.name]: script$a,
        [script$b.name]: script$b,
        [SlotComponent.name]: SlotComponent,
        [script.name]: script,
        [script$1.name]: script$1,
        [script$2.name]: script$2
    },
    mixins: [BaseComponentMixin, MatchMediaMixin],
    configField: 'table',
    inheritAttrs: false,
    provide() {
        return {
            $table: this
        };
    },
    emits: [
        'page-change', 'click', 'dblclick', 'contextmenu',
        'check', 'check-all', 'update:checkedRows',
        'select', 'update:selected', 'filters-change', 'details-open', 'details-close', 'update:openedDetailed',
        'mouseenter', 'mouseleave', 'sort', 'sorting-priority-removed',
        'dragstart', 'dragend', 'drop', 'dragleave', 'dragover', 'cell-click',
        'columndragstart', 'columndragend', 'columndrop', 'columndragleave', 'columndragover',
        'update:currentPage'
    ],
    props: {
        /** Table data */
        data: {
            type: Array,
            default: () => []
        },
        /** Table columns */
        columns: {
            type: Array,
            default: () => []
        },
        /** Border to all cells */
        bordered: Boolean,
        /** Whether table is striped */
        striped: Boolean,
        /** Makes the cells narrower */
        narrowed: Boolean,
        /** Rows are highlighted when hovering */
        hoverable: Boolean,
        /** Loading state */
        loading: Boolean,
        /** Allow row details  */
        detailed: Boolean,
        /** Rows can be checked (multiple), checked rows will have a .is-checked class if you want to style */
        checkable: Boolean,
        /** Show check/uncheck all checkbox in table header when checkable */
        headerCheckable: {
            type: Boolean,
            default: true
        },
        /**
         * Position of the checkbox (if checkable is true)
         * @values left, right
         */
        checkboxPosition: {
            type: String,
            default: 'left',
            validator: (value) => {
                return [
                    'left',
                    'right'
                ].indexOf(value) >= 0;
            }
        },
        /** Set which row is selected, use v-model:selected to make it two-way binding */
        selected: Object,
        /** Custom method to verify if a row is selectable, works when is selected. */
        isRowSelectable: {
            type: Function,
            default: () => true
        },
        /** Table can be focused and user can navigate with keyboard arrows (require selected) and rows are highlighted when hovering */
        focusable: Boolean,
        /** Custom method to verify if row is checked, works when is checkable. Useful for backend pagination */
        customIsChecked: Function,
        /** Custom method to verify if a row is checkable, works when is checkable */
        isRowCheckable: {
            type: Function,
            default: () => true
        },
        /** Set which rows are checked, use v-model:checkedRows to make it two-way binding */
        checkedRows: {
            type: Array,
            default: () => []
        },
        /** Rows appears as cards on mobile (collapse rows) */
        mobileCards: {
            type: Boolean,
            default: () => {
                return getValueByPath(getOptions(), 'table.mobileCards', true);
            }
        },
        /** Sets the default sort column and order — e.g. ['first_name', 'desc']	 */
        defaultSort: [String, Array],
        /**
         * Sets the default sort column direction on the first click
         * @values asc, desc
         */
        defaultSortDirection: {
            type: String,
            default: 'asc'
        },
        /** Sets the header sorting icon */
        sortIcon: {
            type: String,
            default: () => { return getValueByPath(getOptions(), 'table.sortIcon', 'arrow-up'); }
        },
        /**
         * Sets the size of the sorting icon
         * @values small, medium, large
         */
        sortIconSize: {
            type: String,
            default: () => { return getValueByPath(getOptions(), 'table.sortIconSize', 'small'); }
        },
        /** Adds pagination to the table */
        paginated: Boolean,
        /** Current page of table data (if paginated), use v-model:currentPage to make it two-way binding */
        currentPage: {
            type: Number,
            default: 1
        },
        /** How many rows per page (if paginated) */
        perPage: {
            type: [Number, String],
            default: () => { return getValueByPath(getOptions(), 'table.perPage', 20); }
        },
        /** Allow chevron icon and column to be visible */
        showDetailIcon: {
            type: Boolean,
            default: true
        },
        /** Icon name of detail action */
        detailIcon: {
            type: String,
            default: 'chevron-right'
        },
        /**
         * Pagination position (if paginated)
         * @values bottom, top, bot
         */
        paginationPosition: {
            type: String,
            default: () => { return getValueByPath(getOptions(), 'table.paginationPosition', 'bottom'); },
            validator: (value) => {
                return [
                    'bottom',
                    'top',
                    'both'
                ].indexOf(value) >= 0;
            }
        },
        /** Columns won't be sorted with Javascript, use with sort event to sort in your backend */
        backendSorting: Boolean,
        /** Columns won't be filtered with Javascript, use with searchable prop to the columns to filter in your backend */
        backendFiltering: Boolean,
        /** Add a class to row based on the return */
        rowClass: {
            type: Function,
            default: () => ''
        },
        /** Allow pre-defined opened details. Ideal to open details via vue-router. (A unique key is required; check detail-key prop) */
        openedDetailed: {
            type: Array,
            default: () => []
        },
        /** Controls the visibility of the trigger that toggles the detailed rows. */
        hasDetailedVisible: {
            type: Function,
            default: () => true
        },
        /** Use a unique key of your data Object when use detailed or opened detailed. (id recommended) */
        detailKey: {
            type: String,
            default: ''
        },
        /** Custom style on details */
        customDetailRow: {
            type: Boolean,
            default: false
        },
        /* Transition name to use when toggling row details. */
        detailTransition: {
            type: String,
            default: ''
        },
        /** Rows won't be paginated with Javascript, use with page-change event to paginate in your backend */
        backendPagination: Boolean,
        /** Total number of table data if backend-pagination is enabled */
        total: {
            type: [Number, String],
            default: 0
        },
        /** Icon pack to use */
        iconPack: String,
        /** Text when nothing is selected */
        mobileSortPlaceholder: String,
        /** Use a unique key of your data Object for each row. Useful if your data prop has dynamic indices. (id recommended) */
        customRowKey: String,
        /** Allows rows to be draggable */
        draggable: {
            type: Boolean,
            default: false
        },
        /** Allows columns to be draggable */
        draggableColumn: {
            type: Boolean,
            default: false
        },
        /** Add a horizontal scrollbar when table is too wide */
        scrollable: Boolean,
        ariaNextLabel: String,
        ariaPreviousLabel: String,
        ariaPageLabel: String,
        ariaCurrentLabel: String,
        /** Show a sticky table header */
        stickyHeader: Boolean,
        /** Table fixed height */
        height: [Number, String],
        /** Add a native event to filter */
        filtersEvent: {
            type: String,
            default: ''
        },
        /** Filtering debounce time (in milliseconds) */
        debounceSearch: Number,
        /** Show header */
        showHeader: {
            type: Boolean,
            default: () => { return getValueByPath(getOptions(), 'table.showHeader', true); }
        },
        /** Make the checkbox column sticky when checkable */
        stickyCheckbox: {
            type: Boolean,
            default: false
        },
        /** Rounded pagination if paginated */
        paginationRounded: Boolean,
        rootClass: [String, Function, Array],
        tableClass: [String, Function, Array],
        wrapperClass: [String, Function, Array],
        footerClass: [String, Function, Array],
        emptyClass: [String, Function, Array],
        detailedClass: [String, Function, Array],
        borderedClass: [String, Function, Array],
        stripedClass: [String, Function, Array],
        narrowedClass: [String, Function, Array],
        hoverableClass: [String, Function, Array],
        thClass: [String, Function, Array],
        tdClass: [String, Function, Array],
        thPositionClass: [String, Function, Array],
        thStickyClass: [String, Function, Array],
        thCheckboxClass: [String, Function, Array],
        thCurrentSortClass: [String, Function, Array],
        thSortableClass: [String, Function, Array],
        thUnselectableClass: [String, Function, Array],
        thSortIconClass: [String, Function, Array],
        thDetailedClass: [String, Function, Array],
        tdPositionClass: [String, Function, Array],
        tdStickyClass: [String, Function, Array],
        tdCheckboxClass: [String, Function, Array],
        tdDetailedChevronClass: [String, Function, Array],
        trSelectedClass: [String, Function, Array],
        stickyHeaderClass: [String, Function, Array],
        scrollableClass: [String, Function, Array],
        mobileSortClass: [String, Function, Array],
        paginationWrapperClass: [String, Function, Array],
        mobileClass: [String, Function, Array],
        thSubheadingClass: [String, Function, Array]
    },
    data() {
        return {
            visibleDetailRows: this.openedDetailed,
            newData: this.data,
            newDataTotal: this.backendPagination ? this.total : this.data.length,
            newCheckedRows: [...this.checkedRows],
            lastCheckedRowIndex: null,
            newCurrentPage: this.currentPage,
            currentSortColumn: {},
            isAsc: true,
            filters: {},
            defaultSlots: [],
            firstTimeSort: true,
            sequence: 1,
            isDraggingRow: false,
            isDraggingColumn: false
        };
    },
    mounted() {
        this.$nextTick(() => {
            this.checkSort();
        });
    },
    computed: {
        rootClasses() {
            return [
                this.computedClass('rootClass', 'o-table__root'),
                { [this.computedClass('mobileClass', 'o-table__wrapper--mobile')]: this.isMobile }
            ];
        },
        tableClasses() {
            return [
                this.computedClass('tableClass', 'o-table'),
                { [this.computedClass('borderedClass', 'o-table--bordered')]: this.bordered },
                { [this.computedClass('stripedClass', 'o-table--striped')]: this.striped },
                { [this.computedClass('narrowedClass', 'o-table--narrowed')]: this.narrowed },
                { [this.computedClass('hoverableClass', 'o-table--hoverable')]: ((this.hoverable || this.focusable) && this.visibleData.length) },
                { [this.computedClass('emptyClass', 'o-table--table__empty')]: !this.visibleData.length }
            ];
        },
        tableWrapperClasses() {
            return [
                this.computedClass('wrapperClass', 'o-table__wrapper'),
                { [this.computedClass('stickyHeaderClass', 'o-table__wrapper--sticky-header')]: this.stickyHeader },
                { [this.computedClass('scrollableClass', 'o-table__wrapper--scrollable')]: this.isScrollable },
                { [this.computedClass('mobileClass', 'o-table__wrapper--mobile')]: this.isMobile },
            ];
        },
        footerClasses() {
            return [
                this.computedClass('footerClass', 'o-table__footer')
            ];
        },
        thBaseClasses() {
            return [
                this.computedClass('thClass', 'o-table__th')
            ];
        },
        tdBaseClasses() {
            return [
                this.computedClass('tdClass', 'o-table__td')
            ];
        },
        thCheckboxClasses() {
            return [
                ...this.thBaseClasses,
                this.computedClass('thCheckboxClass', 'o-table__th-checkbox')
            ];
        },
        thDetailedClasses() {
            return [
                ...this.thBaseClasses,
                this.computedClass('thDetailedClass', 'o-table__th--detailed')
            ];
        },
        thSubheadingClasses() {
            return [
                ...this.thBaseClasses,
                this.computedClass('thSubheadingClass', 'o-table__th')
            ];
        },
        tdCheckboxClasses() {
            return [
                ...this.tdBaseClasses,
                this.computedClass('tdCheckboxClass', 'o-table__td-checkbox'),
                ...this.thStickyClasses({ sticky: this.stickyCheckbox })
            ];
        },
        detailedClasses() {
            return [
                this.computedClass('detailedClass', 'o-table__detail')
            ];
        },
        tdDetailedChevronClasses() {
            return [
                ...this.tdBaseClasses,
                this.computedClass('tdDetailedChevronClass', 'o-table__td-chevron')
            ];
        },
        mobileSortClasses() {
            return [
                this.computedClass('mobileSortClass', 'o-table__mobile-sort')
            ];
        },
        paginationWrapperClasses() {
            return [
                this.computedClass('paginationWrapperClass', 'o-table__pagination')
            ];
        },
        tableWrapperStyle() {
            return {
                height: toCssDimension(this.height)
            };
        },
        /**
        * Splitted data based on the pagination.
        */
        visibleData() {
            if (!this.paginated)
                return this.newData;
            const currentPage = this.newCurrentPage;
            const perPage = this.perPage;
            if (this.newData.length <= perPage) {
                return this.newData;
            }
            else {
                const start = (currentPage - 1) * perPage;
                const end = start + parseInt(perPage, 10);
                return this.newData.slice(start, end);
            }
        },
        visibleColumns() {
            if (!this.newColumns)
                return this.newColumns;
            return this.newColumns.filter((column) => {
                return column.visible || column.visible === undefined;
            });
        },
        /**
        * Check if all rows in the page are checked.
        */
        isAllChecked() {
            const validVisibleData = this.visibleData.filter((row) => this.isRowCheckable(row));
            if (validVisibleData.length === 0)
                return false;
            const isAllChecked = validVisibleData.some((currentVisibleRow) => {
                return indexOf(this.newCheckedRows, currentVisibleRow, this.customIsChecked) < 0;
            });
            return !isAllChecked;
        },
        /**
        * Check if all rows in the page are checkable.
        */
        isAllUncheckable() {
            const validVisibleData = this.visibleData.filter((row) => this.isRowCheckable(row));
            return validVisibleData.length === 0;
        },
        /**
        * Check if has any sortable column.
        */
        hasSortablenewColumns() {
            return this.newColumns.some((column) => {
                return column.sortable;
            });
        },
        /**
        * Check if has any searchable column.
        */
        hasSearchablenewColumns() {
            return this.newColumns.some((column) => {
                return column.searchable;
            });
        },
        /**
        * Return total column count based if it's checkable or expanded
        */
        columnCount() {
            let count = this.visibleColumns.length;
            count += this.checkable ? 1 : 0;
            count += (this.detailed && this.showDetailIcon) ? 1 : 0;
            return count;
        },
        /**
        * return if detailed row tabled
        * will be with chevron column & icon or not
        */
        showDetailRowIcon() {
            return this.detailed && this.showDetailIcon;
        },
        /**
        * return if scrollable table
        */
        isScrollable() {
            if (this.scrollable)
                return true;
            if (!this.newColumns)
                return false;
            return this.newColumns.some((column) => {
                return column.sticky;
            });
        },
        newColumns() {
            if (this.columns && this.columns.length) {
                return this.columns.map((column) => {
                    const vnode = createVNode(script$1, column, (props) => {
                        const vnode = h('span', {}, getValueByPath(props.row, column.field));
                        return [vnode];
                    });
                    return createApp(vnode)
                        .provide('$table', this)
                        .mount(document.createElement('div'));
                });
            }
            return this.defaultSlots;
        },
        isMobile() {
            return this.mobileCards && this.isMatchMedia;
        },
        hasCustomSubheadings() {
            if (this.$slots.subheading)
                return true;
            return this.newColumns.some((column) => {
                return column.subheading || (column.$slots.subheading);
            });
        },
        canDragRow() {
            return this.draggable && !this.isDraggingColumn;
        },
        canDragColumn() {
            return this.draggableColumn && !this.isDraggingRow;
        }
    },
    watch: {
        /**
        * When data prop change:
        *   1. Update internal value.
        *   2. Filter data if it's not backend-filtered.
        *   3. Sort again if it's not backend-sorted.
        *   4. Set new total if it's not backend-paginated.
        */
        data: {
            handler(value) {
                this.newData = value;
                if (!this.backendFiltering) {
                    this.newData = value.filter((row) => this.isRowFiltered(row));
                }
                if (!this.backendSorting) {
                    this.sort(this.currentSortColumn, true);
                }
                if (!this.backendPagination) {
                    this.newDataTotal = this.newData.length;
                }
            },
            deep: true,
        },
        visibleColumns: {
            handler() {
                this.processTdAttrs();
            }
        },
        visibleData: {
            handler() {
                this.processTdAttrs();
            }
        },
        /**
        * When Pagination total change, update internal total
        * only if it's backend-paginated.
        */
        total(newTotal) {
            if (!this.backendPagination)
                return;
            this.newDataTotal = newTotal;
        },
        currentPage(newValue) {
            this.newCurrentPage = newValue;
        },
        /**
        * When checkedRows prop change, update internal value without
        * mutating original data.
        */
        checkedRows: {
            handler(rows) {
                this.newCheckedRows = [...rows];
            },
            deep: true,
        },
        debounceSearch: {
            handler(value) {
                this.debouncedHandleFiltersChange = debounce(this.handleFiltersChange, value);
            },
            immediate: true
        },
        filters: {
            handler(value) {
                if (this.debounceSearch) {
                    this.debouncedHandleFiltersChange(value);
                }
                else {
                    this.handleFiltersChange(value);
                }
            },
            deep: true,
        },
        /**
        * When the user wants to control the detailed rows via props.
        * Or wants to open the details of certain row with the router for example.
        */
        openedDetailed(expandedRows) {
            this.visibleDetailRows = expandedRows;
        },
        newCurrentPage(newVal) {
            this.$emit('update:currentPage', newVal);
        }
    },
    methods: {
        thClasses(column) {
            return [
                ...this.thBaseClasses,
                ...this.thStickyClasses(column),
                { [this.computedClass('thCurrentSortClass', 'o-table__th-current-sort')]: (this.currentSortColumn === column) },
                { [this.computedClass('thSortableClass', 'o-table__th--sortable')]: column.sortable },
                { [this.computedClass('thUnselectableClass', 'o-table__th--unselectable')]: column.isHeaderUnselectable },
                { [this.computedClass('thPositionClass', 'o-table__th--', column.position)]: column.position },
            ];
        },
        thStickyClasses(column) {
            return [
                { [this.computedClass('thStickyClass', 'o-table__th--sticky')]: column.sticky }
            ];
        },
        rowClasses(row, index) {
            return [
                this.rowClass(row, index),
                { [this.computedClass('trSelectedClass', 'o-table__tr--selected')]: this.isRowSelected(row, this.selected) }
            ];
        },
        thSortIconClasses() {
            return [
                this.computedClass('thSortIconClass', 'o-table__th__sort-icon'),
            ];
        },
        tdClasses(row, column) {
            return [
                ...this.tdBaseClasses,
                { [this.computedClass('tdPositionClass', 'o-table__td--', column.position)]: column.position },
                { [this.computedClass('tdStickyClass', 'o-table__td--sticky')]: column.sticky }
            ];
        },
        onFiltersEvent(event) {
            this.$emit(`filters-event-${this.filtersEvent}`, { event, filters: this.filters });
        },
        handleFiltersChange(value) {
            if (this.backendFiltering) {
                this.$emit('filters-change', value);
            }
            else {
                this.newData = this.data.filter((row) => this.isRowFiltered(row));
                if (!this.backendPagination) {
                    this.newDataTotal = this.newData.length;
                }
                if (!this.backendSorting) {
                    if (Object.keys(this.currentSortColumn).length > 0) {
                        this.doSortSingleColumn(this.currentSortColumn);
                    }
                }
            }
        },
        /**
        * Sort an array by key without mutating original data.
        * Call the user sort function if it was passed.
        */
        sortBy(array, key, fn, isAsc) {
            let sorted = [];
            // Sorting without mutating original data
            if (fn && typeof fn === 'function') {
                sorted = [...array].sort((a, b) => fn(a, b, isAsc));
            }
            else {
                sorted = [...array].sort((a, b) => {
                    // Get nested values from objects
                    let newA = getValueByPath(a, key);
                    let newB = getValueByPath(b, key);
                    // sort boolean type
                    if (typeof newA === 'boolean' && typeof newB === 'boolean') {
                        return isAsc ? newA > newB ? 1 : -1 : newA > newB ? -1 : 1;
                    }
                    if (!newA && newA !== 0)
                        return 1;
                    if (!newB && newB !== 0)
                        return -1;
                    if (newA === newB)
                        return 0;
                    newA = (typeof newA === 'string')
                        ? newA.toUpperCase()
                        : newA;
                    newB = (typeof newB === 'string')
                        ? newB.toUpperCase()
                        : newB;
                    return isAsc
                        ? newA > newB ? 1 : -1
                        : newA > newB ? -1 : 1;
                });
            }
            return sorted;
        },
        /**
        * Sort the column.
        * Toggle current direction on column if it's sortable
        * and not just updating the prop.
        */
        sort(column, updatingData = false, event = null) {
            if (!column || !column.sortable)
                return;
            if (!updatingData) {
                this.isAsc = column === this.currentSortColumn
                    ? !this.isAsc
                    : (this.defaultSortDirection.toLowerCase() !== 'desc');
            }
            if (!this.firstTimeSort) {
                /**
                 * @property {string} field column field
                 * @property {boolean} direction 'asc' or 'desc'
                 * @property {Event} event native event
                */
                this.$emit('sort', column.field, this.isAsc ? 'asc' : 'desc', event);
            }
            if (!this.backendSorting) {
                this.doSortSingleColumn(column);
            }
            this.currentSortColumn = column;
        },
        doSortSingleColumn(column) {
            this.newData = this.sortBy(this.newData, column.field, column.customSort, this.isAsc);
        },
        isRowSelected(row, selected) {
            if (!selected) {
                return false;
            }
            if (this.customRowKey) {
                return row[this.customRowKey] === selected[this.customRowKey];
            }
            return row === selected;
        },
        /**
        * Check if the row is checked (is added to the array).
        */
        isRowChecked(row) {
            return indexOf(this.newCheckedRows, row, this.customIsChecked) >= 0;
        },
        /**
        * Remove a checked row from the array.
        */
        removeCheckedRow(row) {
            const index = indexOf(this.newCheckedRows, row, this.customIsChecked);
            if (index >= 0) {
                this.newCheckedRows.splice(index, 1);
            }
        },
        /**
        * Header checkbox click listener.
        * Add or remove all rows in current page.
        */
        checkAll() {
            const isAllChecked = this.isAllChecked;
            this.visibleData.forEach((currentRow) => {
                if (this.isRowCheckable(currentRow)) {
                    this.removeCheckedRow(currentRow);
                }
                if (!isAllChecked) {
                    if (this.isRowCheckable(currentRow)) {
                        this.newCheckedRows.push(currentRow);
                    }
                }
            });
            /**
             * @property {Array<Object>} newCheckedRows checked rows
             */
            this.$emit('check', this.newCheckedRows);
            this.$emit('check-all', this.newCheckedRows);
            // Emit checked rows to update user variable
            this.$emit('update:checkedRows', this.newCheckedRows);
        },
        /**
        * Row checkbox click listener.
        */
        checkRow(row, index, event) {
            if (!this.isRowCheckable(row))
                return;
            const lastIndex = this.lastCheckedRowIndex;
            this.lastCheckedRowIndex = index;
            if (event.shiftKey && lastIndex !== null && index !== lastIndex) {
                this.shiftCheckRow(row, index, lastIndex);
            }
            else if (!this.isRowChecked(row)) {
                this.newCheckedRows.push(row);
            }
            else {
                this.removeCheckedRow(row);
            }
            this.$emit('check', this.newCheckedRows, row);
            // Emit checked rows to update user variable
            this.$emit('update:checkedRows', this.newCheckedRows);
        },
        /**
         * Check row when shift is pressed.
         */
        shiftCheckRow(row, index, lastCheckedRowIndex) {
            // Get the subset of the list between the two indicies
            const subset = this.visibleData.slice(Math.min(index, lastCheckedRowIndex), Math.max(index, lastCheckedRowIndex) + 1);
            // Determine the operation based on the state of the clicked checkbox
            const shouldCheck = !this.isRowChecked(row);
            subset.forEach((item) => {
                this.removeCheckedRow(item);
                if (shouldCheck && this.isRowCheckable(item)) {
                    this.newCheckedRows.push(item);
                }
            });
        },
        /**
        * Row click listener.
        * Emit all necessary events.
        */
        selectRow(row, index) {
            /**
             * @property {Object} row clicked row
             * @property {number} index index of clicked row
             */
            this.$emit('click', row, index);
            if (this.selected === row)
                return;
            if (!this.isRowSelectable(row))
                return;
            // Emit new and old row
            /**
             * @property {Object} row selected row
             * @property {Array<Object>} selected selected rows
             */
            this.$emit('select', row, this.selected);
            // Emit new row to update user variable
            this.$emit('update:selected', row);
        },
        /**
        * Toggle to show/hide details slot
        */
        toggleDetails(obj) {
            const found = this.isVisibleDetailRow(obj);
            if (found) {
                this.closeDetailRow(obj);
                this.$emit('details-close', obj);
            }
            else {
                this.openDetailRow(obj);
                this.$emit('details-open', obj);
            }
            // Syncs the detailed rows with the parent component
            this.$emit('update:openedDetailed', this.visibleDetailRows);
        },
        openDetailRow(obj) {
            const index = this.handleDetailKey(obj);
            this.visibleDetailRows.push(index);
        },
        closeDetailRow(obj) {
            const index = this.handleDetailKey(obj);
            const i = this.visibleDetailRows.indexOf(index);
            if (i >= 0) {
                this.visibleDetailRows.splice(i, 1);
            }
        },
        isVisibleDetailRow(obj) {
            const index = this.handleDetailKey(obj);
            return this.visibleDetailRows.indexOf(index) >= 0;
        },
        isActiveDetailRow(row) {
            return this.detailed && !this.customDetailRow && this.isVisibleDetailRow(row);
        },
        isActiveCustomDetailRow(row) {
            return this.detailed && this.customDetailRow && this.isVisibleDetailRow(row);
        },
        isRowFiltered(row) {
            for (const key in this.filters) {
                if (!this.filters[key])
                    continue;
                const input = this.filters[key];
                const column = this.newColumns.filter((c) => c.field === key)[0];
                if (column && column.customSearch && typeof column.customSearch === 'function') {
                    if (!column.customSearch(row, input))
                        return false;
                }
                else {
                    const value = getValueByPath(row, key);
                    if (value == null)
                        return false;
                    if (Number.isInteger(value)) {
                        if (value !== Number(input))
                            return false;
                    }
                    else {
                        const re = new RegExp(escapeRegExpChars(input), 'i');
                        if (Array.isArray(value)) {
                            const valid = value.some((val) => re.test(removeDiacriticsFromString(val)) || re.test(val));
                            if (!valid)
                                return false;
                        }
                        else {
                            if (!re.test(removeDiacriticsFromString(value)) && !re.test(value)) {
                                return false;
                            }
                        }
                    }
                }
            }
            return true;
        },
        /**
        * When the detailKey is defined we use the object[detailKey] as index.
        * If not, use the object reference by default.
        */
        handleDetailKey(index) {
            const key = this.detailKey;
            return !key.length || !index
                ? index
                : index[key];
        },
        /**
        * Call initSort only first time (For example async data).
        */
        checkSort() {
            if (this.newColumns.length && this.firstTimeSort) {
                this.initSort();
                this.firstTimeSort = false;
            }
            else if (this.newColumns.length) {
                if (Object.keys(this.currentSortColumn).length > 0) {
                    for (let i = 0; i < this.newColumns.length; i++) {
                        if (this.newColumns[i].field === this.currentSortColumn.field) {
                            this.currentSortColumn = this.newColumns[i];
                            break;
                        }
                    }
                }
            }
        },
        /**
        * Check if footer slot has custom content.
        */
        hasCustomFooterSlot() {
            if (this.$slots.footer) {
                const footer = this.$slots.footer();
                if (footer.length > 1)
                    return true;
                const tag = footer[0].tag;
                if (tag !== 'th' && tag !== 'td')
                    return false;
            }
            return true;
        },
        /**
        * Table arrow keys listener, change selection.
        */
        pressedArrow(pos) {
            if (!this.visibleData.length)
                return;
            let index = this.visibleData.indexOf(this.selected) + pos;
            // Prevent from going up from first and down from last
            index = index < 0
                ? 0
                : index > this.visibleData.length - 1
                    ? this.visibleData.length - 1
                    : index;
            const row = this.visibleData[index];
            if (!this.isRowSelectable(row)) {
                let newIndex = null;
                if (pos > 0) {
                    for (let i = index; i < this.visibleData.length && newIndex === null; i++) {
                        if (this.isRowSelectable(this.visibleData[i]))
                            newIndex = i;
                    }
                }
                else {
                    for (let i = index; i >= 0 && newIndex === null; i--) {
                        if (this.isRowSelectable(this.visibleData[i]))
                            newIndex = i;
                    }
                }
                if (newIndex >= 0) {
                    this.selectRow(this.visibleData[newIndex]);
                }
            }
            else {
                this.selectRow(row);
            }
        },
        /**
        * Focus table element if has selected prop.
        */
        focus() {
            if (!this.focusable)
                return;
            this.$el.querySelector('table').focus();
        },
        /**
        * Initial sorted column based on the default-sort prop.
        */
        initSort() {
            if (!this.defaultSort)
                return;
            let sortField = '';
            let sortDirection = this.defaultSortDirection;
            if (Array.isArray(this.defaultSort)) {
                sortField = this.defaultSort[0];
                if (this.defaultSort[1]) {
                    sortDirection = this.defaultSort[1];
                }
            }
            else {
                sortField = this.defaultSort;
            }
            const sortColumn = this.newColumns.filter((column) => (column.field === sortField))[0];
            if (sortColumn) {
                this.isAsc = sortDirection.toLowerCase() !== 'desc';
                this.sort(sortColumn, true);
            }
        },
        /**
        * Emits drag start event
        */
        handleDragStart(event, row, index) {
            if (!this.draggable)
                return;
            this.$emit('dragstart', { event, row, index });
        },
        /**
        * Emits drag leave event
        */
        handleDragEnd(event, row, index) {
            if (!this.draggable)
                return;
            this.$emit('dragend', { event, row, index });
        },
        /**
        * Emits drop event
        */
        handleDrop(event, row, index) {
            if (!this.draggable)
                return;
            this.$emit('drop', { event, row, index });
        },
        /**
        * Emits drag over event
        */
        handleDragOver(event, row, index) {
            if (!this.draggable)
                return;
            this.$emit('dragover', { event, row, index });
        },
        /**
        * Emits drag leave event
        */
        handleDragLeave(event, row, index) {
            if (!this.draggable)
                return;
            this.$emit('dragleave', { event, row, index });
        },
        /**
        * Emits drag start event (column)
        */
        handleColumnDragStart(event, column, index) {
            if (!this.canDragColumn)
                return;
            this.isDraggingColumn = true;
            this.$emit('columndragstart', { event, column, index });
        },
        /**
        * Emits drag leave event (column)
        */
        handleColumnDragEnd(event, column, index) {
            if (!this.canDragColumn)
                return;
            this.isDraggingColumn = false;
            this.$emit('columndragend', { event, column, index });
        },
        /**
        * Emits drop event (column)
        */
        handleColumnDrop(event, column, index) {
            if (!this.canDragColumn)
                return;
            this.$emit('columndrop', { event, column, index });
        },
        /**
        * Emits drag over event (column)
        */
        handleColumnDragOver(event, column, index) {
            if (!this.canDragColumn)
                return;
            this.$emit('columndragover', { event, column, index });
        },
        /**
        * Emits drag leave event (column)
        */
        handleColumnDragLeave(event, column, index) {
            if (!this.canDragColumn)
                return;
            this.$emit('columndragleave', { event, column, index });
        },
        emitEventForRow(eventName, event, row) {
            return this.$attrs[eventName] ? this.$emit(eventName, row, event) : null;
        },
        processTdAttrs() {
            if (this.visibleColumns.length && this.visibleData.length) {
                for (let i = 0; i < this.visibleColumns.length; i++) {
                    const col = this.visibleColumns[i];
                    if (typeof col.tdAttrs !== 'undefined') {
                        this.visibleData.forEach((data, index) => {
                            col.tdAttrsData[index] = col.tdAttrs(data, col);
                        });
                    }
                }
            }
        },
        _addColumn(column) {
            this.defaultSlots.push(column);
            const slot = this.$refs['slot'];
            if (slot && slot.children) {
                this.$nextTick(() => {
                    const ids = this.defaultSlots.map(it => `[data-id="${it.newKey}"]`).join(',');
                    const sortedIds = Array.from(slot.querySelectorAll(ids)).map((el) => el.getAttribute('data-id'));
                    this.defaultSlots = this.defaultSlots.sort((a, b) => sortedIds.indexOf(`${a.newKey}`) - sortedIds.indexOf(`${b.newKey}`));
                });
            }
        },
        _removeColumn(column) {
            this.defaultSlots = this.defaultSlots.filter(d => d.newKey !== column.newKey);
        },
        _nextSequence() {
            return this.sequence++;
        }
    }
});

const _hoisted_1$1 = {
  ref: "slot",
  style: {
    "display": "none"
  }
};
const _hoisted_2 = {
  key: 0
};
const _hoisted_3 = {
  key: 1
};
const _hoisted_4 = {
  key: 1
};
const _hoisted_5 = {
  key: 0
};
const _hoisted_6 = {
  key: 1
};
const _hoisted_7 = {
  key: 2
};
const _hoisted_8 = {
  key: 1
};
const _hoisted_9 = {
  key: 1
};
const _hoisted_10 = {
  key: 2
};
const _hoisted_11 = {
  key: 0
};
const _hoisted_12 = {
  key: 2
};
function render$2(_ctx, _cache, $props, $setup, $data, $options) {
  const _component_o_table_mobile_sort = resolveComponent("o-table-mobile-sort");

  const _component_o_table_pagination = resolveComponent("o-table-pagination");

  const _component_o_checkbox = resolveComponent("o-checkbox");

  const _component_o_slot_component = resolveComponent("o-slot-component");

  const _component_o_icon = resolveComponent("o-icon");

  const _component_o_input = resolveComponent("o-input");

  const _component_o_loading = resolveComponent("o-loading");

  return openBlock(), createBlock("div", {
    class: _ctx.rootClasses
  }, [createVNode("div", _hoisted_1$1, [renderSlot(_ctx.$slots, "default")], 512
  /* NEED_PATCH */
  ), _ctx.isMobile && _ctx.hasSortablenewColumns ? createVNode(_component_o_table_mobile_sort, {
    key: 0,
    "current-sort-column": _ctx.currentSortColumn,
    columns: _ctx.newColumns,
    placeholder: _ctx.mobileSortPlaceholder,
    "icon-pack": _ctx.iconPack,
    "sort-icon": _ctx.sortIcon,
    "sort-icon-size": _ctx.sortIconSize,
    "is-asc": _ctx.isAsc,
    onSort: _cache[1] || (_cache[1] = (column, event) => _ctx.sort(column, null, event))
  }, null, 8
  /* PROPS */
  , ["current-sort-column", "columns", "placeholder", "icon-pack", "sort-icon", "sort-icon-size", "is-asc"]) : createCommentVNode("v-if", true), _ctx.paginated && (_ctx.paginationPosition === 'top' || _ctx.paginationPosition === 'both') ? renderSlot(_ctx.$slots, "pagination", {
    key: 1
  }, () => [createVNode(_component_o_table_pagination, mergeProps(_ctx.$attrs, {
    "per-page": _ctx.perPage,
    paginated: _ctx.paginated,
    total: _ctx.newDataTotal,
    "current-page": _ctx.newCurrentPage,
    "onUpdate:currentPage": _cache[2] || (_cache[2] = $event => _ctx.newCurrentPage = $event),
    "root-class": _ctx.paginationWrapperClasses,
    "icon-pack": _ctx.iconPack,
    rounded: _ctx.paginationRounded,
    "onPage-change": _cache[3] || (_cache[3] = event => _ctx.$emit('page-change', event)),
    "aria-next-label": _ctx.ariaNextLabel,
    "aria-previous-label": _ctx.ariaPreviousLabel,
    "aria-page-label": _ctx.ariaPageLabel,
    "aria-current-label": _ctx.ariaCurrentLabel
  }), {
    default: withCtx(() => [renderSlot(_ctx.$slots, "top-left")]),
    _: 3
  }, 16
  /* FULL_PROPS */
  , ["per-page", "paginated", "total", "current-page", "root-class", "icon-pack", "rounded", "aria-next-label", "aria-previous-label", "aria-page-label", "aria-current-label"])]) : createCommentVNode("v-if", true), createVNode("div", {
    class: _ctx.tableWrapperClasses,
    style: _ctx.tableWrapperStyle
  }, [createVNode("table", {
    class: _ctx.tableClasses,
    tabindex: !_ctx.focusable ? false : 0,
    onKeydown: [_cache[4] || (_cache[4] = withKeys(withModifiers($event => _ctx.pressedArrow(-1), ["self", "prevent"]), ["up"])), _cache[5] || (_cache[5] = withKeys(withModifiers($event => _ctx.pressedArrow(1), ["self", "prevent"]), ["down"]))]
  }, [_ctx.$slots.caption ? (openBlock(), createBlock("caption", _hoisted_2, [renderSlot(_ctx.$slots, "caption")])) : createCommentVNode("v-if", true), _ctx.newColumns.length && _ctx.showHeader ? (openBlock(), createBlock("thead", _hoisted_3, [createVNode("tr", null, [_ctx.showDetailRowIcon ? (openBlock(), createBlock("th", {
    key: 0,
    class: _ctx.thDetailedClasses
  }, null, 2
  /* CLASS */
  )) : createCommentVNode("v-if", true), _ctx.checkable && _ctx.checkboxPosition === 'left' ? (openBlock(), createBlock("th", {
    key: 1,
    class: _ctx.thCheckboxClasses
  }, [_ctx.headerCheckable ? createVNode(_component_o_checkbox, {
    key: 0,
    autocomplete: "off",
    modelValue: _ctx.isAllChecked,
    disabled: _ctx.isAllUncheckable,
    "onUpdate:modelValue": _ctx.checkAll
  }, null, 8
  /* PROPS */
  , ["modelValue", "disabled", "onUpdate:modelValue"]) : createCommentVNode("v-if", true)], 2
  /* CLASS */
  )) : createCommentVNode("v-if", true), (openBlock(true), createBlock(Fragment, null, renderList(_ctx.visibleColumns, (column, index) => {
    return openBlock(), createBlock("th", mergeProps({
      key: column.newKey + ':' + index + 'header'
    }, column.thAttrsData, {
      class: _ctx.thClasses(column),
      style: _ctx.isMobile ? {} : column.style,
      onClick: withModifiers($event => _ctx.sort(column, null, $event), ["stop"]),
      draggable: _ctx.canDragColumn,
      onDragstart: $event => _ctx.handleColumnDragStart($event, column, index),
      onDragend: $event => _ctx.handleColumnDragEnd($event, column, index),
      onDrop: $event => _ctx.handleColumnDrop($event, column, index),
      onDragover: $event => _ctx.handleColumnDragOver($event, column, index),
      onDragleave: $event => _ctx.handleColumnDragLeave($event, column, index)
    }), [column.hasHeaderSlot ? createVNode(_component_o_slot_component, {
      key: 0,
      component: column,
      name: "header",
      tag: "span",
      props: {
        column,
        index
      }
    }, null, 8
    /* PROPS */
    , ["component", "props"]) : (openBlock(), createBlock("span", _hoisted_4, [createTextVNode(toDisplayString(column.label) + " ", 1
    /* TEXT */
    ), withDirectives(createVNode("span", {
      class: _ctx.thSortIconClasses()
    }, [createVNode(_component_o_icon, {
      icon: _ctx.sortIcon,
      pack: _ctx.iconPack,
      both: "",
      size: _ctx.sortIconSize,
      rotation: !_ctx.isAsc ? 180 : 0
    }, null, 8
    /* PROPS */
    , ["icon", "pack", "size", "rotation"])], 2
    /* CLASS */
    ), [[vShow, column.sortable && _ctx.currentSortColumn === column]])]))], 16
    /* FULL_PROPS */
    , ["onClick", "draggable", "onDragstart", "onDragend", "onDrop", "onDragover", "onDragleave"]);
  }), 128
  /* KEYED_FRAGMENT */
  )), _ctx.checkable && _ctx.checkboxPosition === 'right' ? (openBlock(), createBlock("th", {
    key: 2,
    class: _ctx.thCheckboxClasses
  }, [_ctx.headerCheckable ? createVNode(_component_o_checkbox, {
    key: 0,
    autocomplete: "off",
    modelValue: _ctx.isAllChecked,
    disabled: _ctx.isAllUncheckable,
    "onUpdate:modelValue": _ctx.checkAll
  }, null, 8
  /* PROPS */
  , ["modelValue", "disabled", "onUpdate:modelValue"]) : createCommentVNode("v-if", true)], 2
  /* CLASS */
  )) : createCommentVNode("v-if", true)]), _ctx.hasSearchablenewColumns ? (openBlock(), createBlock("tr", _hoisted_5, [_ctx.showDetailRowIcon ? (openBlock(), createBlock("th", {
    key: 0,
    class: _ctx.thDetailedClasses
  }, null, 2
  /* CLASS */
  )) : createCommentVNode("v-if", true), _ctx.checkable && _ctx.checkboxPosition === 'left' ? (openBlock(), createBlock("th", _hoisted_6)) : createCommentVNode("v-if", true), (openBlock(true), createBlock(Fragment, null, renderList(_ctx.visibleColumns, (column, index) => {
    return openBlock(), createBlock("th", mergeProps({
      key: column.newKey + ':' + index + 'searchable'
    }, column.thAttrsData, {
      class: _ctx.thClasses(column),
      style: _ctx.isMobile ? {} : column.style
    }), [column.searchable ? (openBlock(), createBlock(Fragment, {
      key: 0
    }, [column.hasSearchableSlot ? createVNode(_component_o_slot_component, {
      key: 0,
      component: column,
      name: "searchable",
      tag: "span",
      props: {
        column,
        filters: _ctx.filters
      }
    }, null, 8
    /* PROPS */
    , ["component", "props"]) : createVNode(_component_o_input, {
      key: 1,
      ["on" + capitalize(_ctx.filtersEvent)]: _ctx.onFiltersEvent,
      modelValue: _ctx.filters[column.field],
      "onUpdate:modelValue": $event => _ctx.filters[column.field] = $event,
      type: column.numeric ? 'number' : 'text'
    }, null, 16
    /* FULL_PROPS */
    , ["modelValue", "onUpdate:modelValue", "type"])], 64
    /* STABLE_FRAGMENT */
    )) : createCommentVNode("v-if", true)], 16
    /* FULL_PROPS */
    );
  }), 128
  /* KEYED_FRAGMENT */
  )), _ctx.checkable && _ctx.checkboxPosition === 'right' ? (openBlock(), createBlock("th", _hoisted_7)) : createCommentVNode("v-if", true)])) : createCommentVNode("v-if", true), _ctx.hasCustomSubheadings ? (openBlock(), createBlock("tr", _hoisted_8, [_ctx.showDetailRowIcon ? (openBlock(), createBlock("th", {
    key: 0,
    class: _ctx.thDetailedClasses
  }, null, 2
  /* CLASS */
  )) : createCommentVNode("v-if", true), _ctx.checkable && _ctx.checkboxPosition === 'left' ? (openBlock(), createBlock("th", _hoisted_9)) : createCommentVNode("v-if", true), (openBlock(true), createBlock(Fragment, null, renderList(_ctx.visibleColumns, (column, index) => {
    return openBlock(), createBlock("th", {
      key: column.newKey + ':' + index + 'subheading',
      style: _ctx.isMobile ? {} : column.style,
      class: _ctx.thSubheadingClasses
    }, [column.$slots && column.$slots.subheading ? createVNode(_component_o_slot_component, {
      key: 0,
      component: column,
      name: "subheading",
      tag: "span",
      props: {
        column,
        index
      }
    }, null, 8
    /* PROPS */
    , ["component", "props"]) : (openBlock(), createBlock(Fragment, {
      key: 1
    }, [createTextVNode(toDisplayString(column.subheading), 1
    /* TEXT */
    )], 64
    /* STABLE_FRAGMENT */
    ))], 6
    /* CLASS, STYLE */
    );
  }), 128
  /* KEYED_FRAGMENT */
  )), _ctx.checkable && _ctx.checkboxPosition === 'right' ? (openBlock(), createBlock("th", _hoisted_10)) : createCommentVNode("v-if", true)])) : createCommentVNode("v-if", true)])) : createCommentVNode("v-if", true), createVNode("tbody", null, [(openBlock(true), createBlock(Fragment, null, renderList(_ctx.visibleData, (row, index) => {
    return openBlock(), createBlock(Fragment, {
      key: this.customRowKey ? row[this.customRowKey] : index
    }, [createVNode("tr", {
      class: _ctx.rowClasses(row, index),
      onClick: $event => _ctx.selectRow(row, index),
      onDblclick: $event => _ctx.$emit('dblclick', row),
      onMouseenter: $event => _ctx.emitEventForRow('mouseenter', $event, row),
      onMouseleave: $event => _ctx.emitEventForRow('mouseleave', $event, row),
      onContextmenu: $event => _ctx.$emit('contextmenu', row, $event),
      draggable: _ctx.canDragRow,
      onDragstart: $event => _ctx.handleDragStart($event, row, index),
      onDragend: $event => _ctx.handleDragEnd($event, row, index),
      onDrop: $event => _ctx.handleDrop($event, row, index),
      onDragover: $event => _ctx.handleDragOver($event, row, index),
      onDragleave: $event => _ctx.handleDragLeave($event, row, index)
    }, [_ctx.showDetailRowIcon ? (openBlock(), createBlock("td", {
      key: 0,
      class: _ctx.tdDetailedChevronClasses
    }, [_ctx.hasDetailedVisible(row) ? createVNode(_component_o_icon, {
      key: 0,
      icon: _ctx.detailIcon,
      pack: _ctx.iconPack,
      rotation: _ctx.isVisibleDetailRow(row) ? 90 : 0,
      role: "button",
      onClick: withModifiers($event => _ctx.toggleDetails(row), ["stop"]),
      clickable: "",
      both: ""
    }, null, 8
    /* PROPS */
    , ["icon", "pack", "rotation", "onClick"]) : createCommentVNode("v-if", true)], 2
    /* CLASS */
    )) : createCommentVNode("v-if", true), _ctx.checkable && _ctx.checkboxPosition === 'left' ? (openBlock(), createBlock("td", {
      key: 1,
      class: _ctx.tdCheckboxClasses
    }, [createVNode(_component_o_checkbox, {
      autocomplete: "off",
      disabled: !_ctx.isRowCheckable(row),
      modelValue: _ctx.isRowChecked(row),
      "onUpdate:modelValue": $event => _ctx.checkRow(row, index, $event)
    }, null, 8
    /* PROPS */
    , ["disabled", "modelValue", "onUpdate:modelValue"])], 2
    /* CLASS */
    )) : createCommentVNode("v-if", true), (openBlock(true), createBlock(Fragment, null, renderList(_ctx.visibleColumns, (column, colindex) => {
      return openBlock(), createBlock(_component_o_slot_component, mergeProps({
        key: column.newKey + index + ':' + colindex
      }, column.tdAttrsData[index], {
        component: column,
        name: "default",
        tag: "td",
        class: _ctx.tdClasses(row, column),
        style: _ctx.isMobile ? {} : column.style,
        "data-label": column.label,
        props: {
          row,
          column,
          index,
          colindex,
          toggleDetails: _ctx.toggleDetails
        },
        onClick: $event => _ctx.$emit('cell-click', row, column, index, colindex, $event)
      }), null, 16
      /* FULL_PROPS */
      , ["component", "class", "style", "data-label", "props", "onClick"]);
    }), 128
    /* KEYED_FRAGMENT */
    )), _ctx.checkable && _ctx.checkboxPosition === 'right' ? (openBlock(), createBlock("td", {
      key: 2,
      class: _ctx.tdCheckboxClasses
    }, [createVNode(_component_o_checkbox, {
      autocomplete: "off",
      disabled: !_ctx.isRowCheckable(row),
      modelValue: _ctx.isRowChecked(row),
      "onUpdate:modelValue": $event => _ctx.checkRow(row, index, $event)
    }, null, 8
    /* PROPS */
    , ["disabled", "modelValue", "onUpdate:modelValue"])], 2
    /* CLASS */
    )) : createCommentVNode("v-if", true)], 42
    /* CLASS, PROPS, HYDRATE_EVENTS */
    , ["onClick", "onDblclick", "onMouseenter", "onMouseleave", "onContextmenu", "draggable", "onDragstart", "onDragend", "onDrop", "onDragover", "onDragleave"]), createVNode(Transition, {
      name: _ctx.detailTransition
    }, {
      default: withCtx(() => [_ctx.isActiveDetailRow(row) ? (openBlock(), createBlock("tr", {
        key: (_ctx.customRowKey ? row[_ctx.customRowKey] : index) + 'detail',
        class: _ctx.detailedClasses
      }, [createVNode("td", {
        colspan: _ctx.columnCount
      }, [renderSlot(_ctx.$slots, "detail", {
        row: row,
        index: index
      })], 8
      /* PROPS */
      , ["colspan"])], 2
      /* CLASS */
      )) : createCommentVNode("v-if", true)]),
      _: 2
    }, 1032
    /* PROPS, DYNAMIC_SLOTS */
    , ["name"]), _ctx.isActiveCustomDetailRow(row) ? renderSlot(_ctx.$slots, "detail", {
      key: 0,
      row: row,
      index: index
    }) : createCommentVNode("v-if", true)], 64
    /* STABLE_FRAGMENT */
    );
  }), 128
  /* KEYED_FRAGMENT */
  )), !_ctx.visibleData.length ? (openBlock(), createBlock("tr", _hoisted_11, [createVNode("td", {
    colspan: _ctx.columnCount
  }, [renderSlot(_ctx.$slots, "empty")], 8
  /* PROPS */
  , ["colspan"])])) : createCommentVNode("v-if", true)]), _ctx.$slots.footer ? (openBlock(), createBlock("tfoot", _hoisted_12, [createVNode("tr", {
    class: _ctx.footerClasses
  }, [_ctx.hasCustomFooterSlot() ? renderSlot(_ctx.$slots, "footer", {
    key: 0
  }) : (openBlock(), createBlock("th", {
    key: 1,
    colspan: _ctx.columnCount
  }, [renderSlot(_ctx.$slots, "footer")], 8
  /* PROPS */
  , ["colspan"]))], 2
  /* CLASS */
  )])) : createCommentVNode("v-if", true)], 42
  /* CLASS, PROPS, HYDRATE_EVENTS */
  , ["tabindex"]), renderSlot(_ctx.$slots, "loading", {}, () => [createVNode(_component_o_loading, {
    "full-page": false,
    active: _ctx.loading
  }, null, 8
  /* PROPS */
  , ["active"])])], 6
  /* CLASS, STYLE */
  ), _ctx.checkable && _ctx.$slots['bottom-left'] || _ctx.paginated && (_ctx.paginationPosition === 'bottom' || _ctx.paginationPosition === 'both') ? renderSlot(_ctx.$slots, "pagination", {
    key: 2
  }, () => [createVNode(_component_o_table_pagination, mergeProps(_ctx.$attrs, {
    "per-page": _ctx.perPage,
    paginated: _ctx.paginated,
    total: _ctx.newDataTotal,
    "current-page": _ctx.newCurrentPage,
    "onUpdate:currentPage": _cache[6] || (_cache[6] = $event => _ctx.newCurrentPage = $event),
    "root-class": _ctx.paginationWrapperClasses,
    "icon-pack": _ctx.iconPack,
    rounded: _ctx.paginationRounded,
    "onPage-change": _cache[7] || (_cache[7] = event => _ctx.$emit('page-change', event)),
    "aria-next-label": _ctx.ariaNextLabel,
    "aria-previous-label": _ctx.ariaPreviousLabel,
    "aria-page-label": _ctx.ariaPageLabel,
    "aria-current-label": _ctx.ariaCurrentLabel
  }), {
    default: withCtx(() => [renderSlot(_ctx.$slots, "bottom-left")]),
    _: 3
  }, 16
  /* FULL_PROPS */
  , ["per-page", "paginated", "total", "current-page", "root-class", "icon-pack", "rounded", "aria-next-label", "aria-previous-label", "aria-page-label", "aria-current-label"])]) : createCommentVNode("v-if", true)], 2
  /* CLASS */
  );
}

script$3.render = render$2;
script$3.__file = "src/components/table/Table.vue";

var index = {
    install(app) {
        registerComponent(app, script$3);
        registerComponent(app, script$1);
    }
};

export default index;
export { script$3 as OTable, script$1 as OTableColumn };
