// @oruga-ui/oruga-next/dist/cjs/MatchMediaMixin-72cab4f9.js
'use strict';

var vue = require('vue');
var helpers = require('./helpers.js');
var config = require('./config.js');

var MatchMediaMixin = vue.defineComponent({
    props: {
        /**
         * Mobile breakpoint as max-width value
         */
        mobileBreakpoint: String
    },
    data() {
        return {
            matchMediaRef: undefined,
            isMatchMedia: undefined
        };
    },
    methods: {
        onMatchMedia(event) {
            this.isMatchMedia = event.matches;
        }
    },
    created() {
        if (typeof window !== 'undefined') {
            let width = this.mobileBreakpoint;
            if (!width) {
                const config$1 = config.getOptions();
                const defaultWidth = helpers.getValueByPath(config$1, `mobileBreakpoint`, '1023px');
                width = helpers.getValueByPath(config$1, `${this.$options.configField}.mobileBreakpoint`, defaultWidth);
            }
            this.matchMediaRef = window.matchMedia(`(max-width: ${width})`);
            if (this.matchMediaRef) {
                this.isMatchMedia = this.matchMediaRef.matches;
                this.matchMediaRef.addListener(this.onMatchMedia, false);
            }
            else {
                this.isMatchMedia = false;
            }
        }
    },
    beforeUnmount() {
        if (typeof window !== 'undefined') {
            if (this.matchMediaRef) {
                this.matchMediaRef.removeListener(this.checkMatchMedia);
            }
        }
    }
});

exports.MatchMediaMixin = MatchMediaMixin;
